src/Handler/Backend/Init/InitBackendHandler.php

```diff
@@ -49,25 +70,156 @@ public function __invoke(
+                $folderName,
+                self::EXPECTED_FOLDER_PERMISSIONS,
             );
+        } catch (ApiException $e) {
+            $exceptions[] = new Exception(sprintf(
+                'Cannot get permissions for folder "%s" expected permission "%s" was not probably assigned. "%s"',
+                $folderName,
+                implode(', ', self::EXPECTED_FOLDER_PERMISSIONS),
+                $e->getReason(),
+            ));
         } finally {
             $foldersClient->close();
         }
+        $this->assertPermissions($exceptions, self::EXPECTED_FOLDER_PERMISSIONS, $folderPermissions);
+
+        // check root project roles
+        $projectsClient = $this->clientManager->getProjectClient($credentials);
+        /** @var array<string, string> $principal */
+        $principal = (array) json_decode($credentials->getPrincipal(), true, 512, JSON_THROW_ON_ERROR);
+        $projectNameFormatted = $projectsClient::projectName($principal['project_id']);
+
+        $policies = null;
+        try {
+            $policies = $projectsClient->getIamPolicy($projectNameFormatted);
+        } catch (ApiException $e) {
+            $exceptions[] = new Exception(sprintf(
+                'Cannot get roles for project "%s" expected roles "%s" was not probably assigned. "%s"',
+                $projectNameFormatted,
+                implode(', ', self::EXPECTED_PROJECT_ROLES),
+                $e->getReason(),
+            ));
+        } finally {
+            $projectsClient->close();
+        }
+        $roles = $this->getRolesFromPolicy($policies);
+        $this->assertRoles($exceptions, self::EXPECTED_PROJECT_ROLES, $roles);
 
-        if (count($folderPermissions->getPermissions()) === 0) {
-            throw new Exception(sprintf(
-                'Missing rights "%s for service account.',
-                IAmPermissions::RESOURCE_MANAGER_PROJECTS_CREATE,
+        // check billing account roles
+        $billingClient = $this->clientManager->getBillingClient($credentials);
+        $billingInfo = null;
+        try {
+            $billingInfo = $billingClient->getProjectBillingInfo($projectNameFormatted);
+        } catch (ApiException $e) {
+            $exceptions[] = new Exception(sprintf(
+                'Cannot fetch project "%s" billing info. "%s"',
+                $projectNameFormatted,
+                $e->getReason(),
             ));
         }
+        if ($billingInfo !== null) {
+            $mainBillingAccount = $billingInfo->getBillingAccountName();
+            try {
+                $policies = $billingClient->getIamPolicy($mainBillingAccount);
+            } catch (ApiException $e) {
+                $exceptions[] = new Exception(sprintf(
+                    'Cannot get roles for billing account "%s" expected roles "%s" was not probably assigned. "%s"',
+                    $mainBillingAccount,
+                    implode(', ', self::EXPECTED_BILLING_ROLES),
+                    $e->getReason(),
```

Comment:
Ted mi to skoncilo s `Cannot get roles for billing account "billingAccounts/01C438-814D25-220303" expected roles "roles/billing.user" was not probably assigned. ""` . Ten `reason` tam asi neni a ty "" vypadaji dost pitomne. Nedame tam neco jako `$e->getReason() ?: 'no explanation provided`

Reply:
mysleno asi vsude